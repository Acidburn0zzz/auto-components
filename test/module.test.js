const { join } = require('path')
const { loadConfig, init } = require('@nuxtjs/module-test-utils')
const chokidar = require('chokidar')
const globby = require('globby')

// Mock globby
jest.mock('globby')
const toGlobbyFiles = arr => arr.map(name => ({ name, path: join('/project/nuxt', name) }))
globby.mockImplementation(() => toGlobbyFiles(['a.js', 'b.global.js', 'c.global.js']))

// Mock chokidar
const chokidarWatchers = []
const mockWatcher = {
  close: jest.fn(),
  on (event, fn) {
    chokidarWatchers.push({ event, fn })
  }
}
chokidar.watch = () => mockWatcher
const callChokidarEvent = eventName => Promise.all(chokidarWatchers.map(w => w.fn(eventName)))

describe('module', () => {
  let nuxt, builder

  beforeAll(async () => {
    nuxt = await init(loadConfig(__dirname, undefined, { dev: true }))
    builder = {
      generateRoutesAndFiles: jest.fn()
    }
    await nuxt.callHook('build:before', builder)
  })

  afterAll(async () => {
    await nuxt.close()
  })

  test('discover components and pass as componentOptions', () => {
    const plugin = nuxt.options.build.templates.find(p => p.dst.includes('global-components'))
    expect(plugin.options).toMatchSnapshot()
  })

  test('watch: no rebuild if no matched files changed', async () => {
    builder.generateRoutesAndFiles.mockClear()
    await callChokidarEvent('add')
    expect(builder.generateRoutesAndFiles).toHaveBeenCalledTimes(0)
  })

  test('watch:rebuild on add/remove', async () => {
    builder.generateRoutesAndFiles.mockClear()
    globby.mockImplementationOnce(() => toGlobbyFiles(['a.js', 'b.global.js']))

    await callChokidarEvent('unlink')
    expect(builder.generateRoutesAndFiles).toHaveBeenCalledTimes(1)
  })

  test('watch:no rebuild on other events', async () => {
    builder.generateRoutesAndFiles.mockClear()
    globby.mockImplementationOnce(() => toGlobbyFiles(['a.js', 'b.global.js']))

    await callChokidarEvent('foo')
    expect(builder.generateRoutesAndFiles).toHaveBeenCalledTimes(0)
  })
})
