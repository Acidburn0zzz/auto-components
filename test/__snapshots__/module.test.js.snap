// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module discover components and pass as componentOptions 1`] = `
Object {
  "components": Array [
    Object {
      "name": "a.js",
      "path": "~/components//project/nuxt/a.js",
    },
    Object {
      "name": "b",
      "path": "~/components//project/nuxt/b.global.js",
    },
    Object {
      "name": "c",
      "path": "~/components//project/nuxt/c.global.js",
    },
  ],
  "ignoreNameDetection": false,
}
`;
